### ISSUE TITLE
Inaccurate Isochrones/SPT's where intersections are sparse

### ISSUE THREAD

**Issue by EddyIonescu** [graphhopper/graphhopper#2377]
**Describe the bug**

One use-case of GH is to create isochrones to show how far one can get. However, these isochrones and SPT's can be fairly imprecise for situations where the time or distance limit is short and the road network is sparse (ie. few intersections like in rural or suburban areas).

**To Reproduce**

- Create an isochrone or SPT on a road where there are no intersections within the time/distance limit. 
<img src="https://user-images.githubusercontent.com/8988797/128793362-b41747ed-d521-4f16-9db1-478bfa49e235.png" width=40% height=40%>
Figure 1 - distance_limit=1100m, profile=foot

<img src="https://user-images.githubusercontent.com/8988797/128793808-c396c82c-7d76-43d4-ade2-8f5819d928e0.png" width=40% height=40%>
Figure 2 - distance_limit=1150m, profile=foot

- Alternatively, create an isochrone or SPT on a road where there are few intersections within the time/distance limit (Figures 3 and 4).
<img src="https://user-images.githubusercontent.com/8988797/128793389-bf77bf29-c2bd-4ffb-b1cd-a3c43309f363.png" height=40% width=40%>
Figure 3 - distance_limit=400, profile=foot

<img src="https://user-images.githubusercontent.com/8988797/128793406-c9dd4ee5-6bc6-433e-be8c-aa5264491561.png" width=40% height=40%>
Figure 4 - distance_limit=400, origin moved very slightly to the west; profile=foot.

Note that the isochrones above are shown exactly as returned by GH, and aren't processed by any other tools.

**Expected behavior**
Tracing through from the origin to the end of each line in the SPT gives you the distance_limit, which would be similar to what you'd get when making a Service Area in ArcMap.

**System Information**

GH 3.2

**Approach to Resolve**

Having looked through the codebase I'm looking for some guidance into what would be required to resolve this issue.

One possible approach I was thinking of is, after having generated the edges in the SPT, extend them along the road away from the origin until the exact distance/time limit is reached. If this ends up degrading performance then making it optional could work for turning it into a PR.

Thanks!

**Comment by michaz** [comment]
> One possible approach I was thinking of is, after having generated the edges in the SPT, extend them along the road away from the origin until the exact distance/time limit is reached. If this ends up degrading performance then making it optional could work for turning it into a PR.

Yes, absolutely. I think this is on the right track.

**Comment by Rikuoja** [comment]
We would be very interested in this implementation as well, our blog post (sorry it's a bit lengthy) contains some issues with the current isochrone implementation and incomplete OSM data: https://www.gispo.fi/en/blog/analyzing-school-accessibility-on-qgis/

**Tl;dr**: Our idea would be to just buffer the accessible network with a constant buffer to create the polygons. Is this too simple, do you believe travel time across terrain should be taken into account too?

That would cause problems when roads are sparse. I believe we cannot assume that all terrain is traversable starting from the road network, since in reality there will be obstacles, some of them insurmountable. 

It would make more sense for us to use a smallish buffer instead of assuming that all points are accessible from the road.

**Comment by michaz** [comment]
@Rikuoja Thanks! Before reading your blog post, let me just quickly say that, in my mind, that's a slightly different issue: How is an (abstract) isochrone _on a road network_ extended to an isochrone in the terrain.

This here is more specifically (and more simply) about making the isochrone on the road network more fine-grained. At the moment, it basically snaps to intersections, leading to inaccuracies when _intersections_ are sparse (rather than _roads_).

The issue you're raising is much broader.

**Comment by Rikuoja** [comment]
@michaz A good point! Indeed, just adding nodes or intersections where they are sparse would go a long way to making the isochrone shape closer to reality, sounds like a good start.

**Comment by econaxis** [comment]
Hey @michaz , took a stab at this over at #2838 based on your approach described above.
